EIPS/eip-6228.md

```diff
@@ -0,0 +1,233 @@
+        bool forbidAll; // emergency case, etc
+    }
+```
+
+### Method callBySignature()
+
+MUST not be used in production. This is an example of EIP callBySignature() function. An update with a tested version will come out soon.
+
+```SOLIDITY
+    function callBySignature(Request calldata reqt) external payable {
+        uint initGasLeft = gasleft();// gas comes first
+        uint gasPrice;
+        assembly {
+            gasPrice := gasprice()
+        }
+        require(!users[reqt.signer].forbidAll, 'forbid all is active');
+        _verifySig(reqt);
+        // check if allowed to use this signature
+        require(block.number > users[reqt.signer].indexLastClaims[reqt.index] + reqt.interval, 'sig cant be valid yet');
+        require(users[reqt.signer].forbiddenSaltsBitMap[reqt.salt / 256] & (1 << reqt.salt % 256) == 0, 'forbidden salt');
+        require(reqt.interval >= 86400);// in seconds, this is to avoid griefing
+        // record state change if any
+        if (reqt.interval > 0) {
+            users[reqt.signer].indexLastClaims[reqt.index] = block.number; // its a subscription, track last execution time
+        } else {
+            _forbidSalt(reqt.signer, reqt.salt);// it was a one time transaction, forbid this salt
+        }
+        // call
+        // if data is 0, subtract keeper reward from user balance here
+        // otherwise data should return amount which is then being split between the user and the keeper
+        bytes memory data;
+        bool success;
+        if (reqt.needsHelper) {
+            (success, data) = I(_ab.helper).handle{value: msg.value}(reqt);
+            require(success == true, 'helper failed');
+        } else {
+            (success, data) = reqt.callee.call{value: msg.value}(abi.encodeWithSelector(reqt.selector, reqt.callData));
+            require(success == true, 'direct call failed');
+        }
+        // take fee and complete
+        uint amount; // compute amount from data bytes
+        assembly {
+            amount := mload(add(data, 0x20))
+        }
+        uint approxGas = (initGasLeft - gasleft() /*+ 21000 */) * gasPrice;
+        uint keeperReward = _calculateAmountInNativeToken(reqt.returnToken, approxGas + reqt.tip);
+        require(amount - keeperReward >= reqt.minOut);
+        if (reqt.returnToken == address(this)) {
+            users[msg.sender].balance += keeperReward;
+            users[reqt.signer].balance += amount - keeperReward;
+        } else {
+            if (reqt.returnToken == address(0)) {
+                (success, data) = payable(msg.sender).call{value: keeperReward}('');
+                require(success);
+                (success, data) = payable(reqt.signer).call{value: amount - keeperReward}('');
+                require(success);
+            } else {
+                I(reqt.returnToken).transfer(msg.sender, keeperReward);
+                I(reqt.returnToken).transfer(reqt.signer, amount - keeperReward);
+            }
+        }
```

Comment:
```suggestion
```

You shouldn't include any implementation in the specification section, only in the Reference Implementation.